### DISCUSSION TITLE
Improve JulianDateTime

### DISCUSSION THREAD

**Discussion by igr** [oblac/jodd-util#14] [General] [answered]
Let's discuss here all the issues and corner cases of the `JulianDateTime`

@cusher let's list everything we may fix with the JulianDateTime.

**Comment by cusher** [comment, marked as answer]
I came up a slightly obnoxious test case for precision. Quick calculation to start off with:
> 1 JD = 86400s
> 0.00000000001 JD = 864ns
> 0.00000000578 JD = 499392ns = 0.499392ms
> 0.00000000579 JD = 500256ns = 0.500256ms

So in theory if we add/subtract `0.00000000578` from any JulianDate that is a whole number of millis, it should not round away, while if we add/subtract `0.00000000579` it should round away.

And so:
```
	@Test
	void bigDecimalToLocalDateTimeDelta() {
		JulianDate jdt = new JulianDate(BigDecimal.valueOf(2457754.4));
		LocalDateTime ldt = LocalDateTime.of(2016, 12, 31, 21, 36, 0, 0);

		assertEquals(ldt, jdt.toLocalDateTime());
		assertEquals(ldt, jdt.sub(0.00000000578).toLocalDateTime());
		assertEquals(ldt, jdt.add(0.00000000578).toLocalDateTime());
		assertEquals(ldt.minus(1, ChronoUnit.MILLIS), jdt.sub(0.00000000579).toLocalDateTime());
		assertEquals(ldt.plus(1, ChronoUnit.MILLIS), jdt.add(0.00000000579).toLocalDateTime());
	}
```
That's the only other thing I can think of.

**Reply by igr** [reply]
I've added the following:

```
final int time_millisecond = (int) Math.round(d_millis - 0.0085);
```

in the method `toLocalDateTime()` and this makes this test green (other tests as well).

Do you think we can commit this @cusher ?

**Reply by cusher** [reply]
@igr This test should already pass, in fact I added an updated version of this check in `testDeltaRounding`. The change I made that got this test passing was removing the `+ 1.0e-10` in `toLocalDateTime:`

https://github.com/oblac/jodd-util/blob/695d909b0de045a2b41aac3ad74635f65f07d4d6/src/main/java/jodd/time/JulianDate.java#L317

https://github.com/oblac/jodd-util/blob/833d87259edb25198a2c4112e39c801938874a81/src/main/java/jodd/time/JulianDate.java#L319

**Reply by igr** [reply]
Got it, I was solving the same issue in a different place! I just didn't pull 🤦‍♂️ (sorry, doing couple of things in the same time:)

We can release now I would say :)
